rust/lance/src/index/vector/ivf/shuffler.rs

```diff
@@ -0,0 +1,199 @@
+        }
+
+        file_writer.finish().await?;
+
+        Ok(())
+    }
+
+    pub async fn count_partition_size(&self) -> Result<Vec<u64>> {
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        let mut partition_sizes = vec![0; self.num_partitions as usize];
+
+        let lance_schema = reader
+            .schema()
+            .project(&[PART_ID_COLUMN])
+            .expect("part id should exist");
+
+        let mut stream = stream::iter(0..reader.num_batches())
+            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, &lance_schema))
+            .buffered(64);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+            let part_ids: &UInt32Array = as_primitive_array(batch.column(0));
+            for part_id in part_ids.values() {
+                partition_sizes[*part_id as usize] += 1;
+            }
+        }
+
+        Ok(partition_sizes)
+    }
+
+    pub async fn shuffle_to_partitions(
+        &self,
+        partition_size: Vec<u64>,
+    ) -> Result<(Vec<Vec<u64>>, Vec<Vec<u8>>)> {
+        let mut row_id_buffers = partition_size
+            .iter()
+            .map(|s| Vec::with_capacity(*s as usize))
+            .collect::<Vec<_>>();
+        let mut pq_code_buffers = partition_size
+            .iter()
+            .map(|s| Vec::with_capacity((*s as usize) * self.pq_width))
+            .collect::<Vec<_>>();
+
+        let object_store = ObjectStore::local();
+        let path = self.output_dir.child(UNSORTED_BUFFER);
+        let reader = FileReader::try_new(&object_store, &path).await?;
+
+        info!("Shuffling into memory");
+        let mut bar = kdam::BarBuilder::default()
+            .total(reader.num_batches())
+            .build()
+            .unwrap();
+
+        let mut stream = stream::iter(0..reader.num_batches())
+            .map(|i| reader.read_batch(i as i32, ReadBatchParams::RangeFull, reader.schema()))
+            .buffered(16);
+
+        while let Some(batch) = stream.next().await {
+            let batch = batch?;
+
+            let row_ids = as_primitive_array::<UInt64Type>(
+                batch
+                    .column_by_name(ROW_ID)
+                    .expect("Row ID column not found"),
+            );
+
+            let part_ids = as_primitive_array::<UInt32Type>(
+                batch
+                    .column_by_name(PART_ID_COLUMN)
+                    .expect("Partition ID column not found"),
+            );
+
+            let pq_codes_arr = batch
+                .column_by_name(PQ_CODE_COLUMN)
+                .expect("PQ Code column not found");
+
+            let pq_codes =
+                as_primitive_array::<UInt8Type>(as_fixed_size_list_array(pq_codes_arr).values());
+
+            let num_sub_vectors = pq_codes.len() / row_ids.len();
+
+            row_ids
```

Comment:
wait, this means all shufflig just happen in memory?

Reply:
yes, we are shuffling into memory

Reply:
This will soon be problematic if we have more than a few billion vectors? 

shuffle v1 was meant to use disk based external shuffling

Reply:
For 1B 768D, I think memory could still handle. We just need to implement external shuffle before that. I wanted to avoid the complexity in this PR. We can do a quick follow up I guess.

Reply:
Ok, lets create an issue and address it soon?